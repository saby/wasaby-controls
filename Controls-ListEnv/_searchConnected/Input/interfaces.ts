import { IContextOptionsValue } from 'Controls/context';
import { ISearchInputContainerOptions } from 'Controls/search';

/**
 * Опции виджета строки поиска
 * @private
 */
export interface ISearchInputOptions extends ISearchInputContainerOptions {
    /**
     * @cfg {string} Идентификатор списка с которым связан виджет поиска.
     */
    storeId: string | number;

    _dataOptionsValue: IContextOptionsValue;
}
