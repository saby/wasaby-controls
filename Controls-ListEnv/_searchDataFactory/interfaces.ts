export interface ISearchState {
    searchValue: string;
    searchInputValue: string;
}

export interface ISearchDataFactoryParams {
    searchValue: string;
}

export interface ISearchLoadResult {
    searchValue?: string;
}
