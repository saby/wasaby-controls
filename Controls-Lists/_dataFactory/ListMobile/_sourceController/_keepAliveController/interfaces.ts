/**
 * Параметры конструктора источника данных
 * @private
 */
export interface TGinniSourceConstructorOptions {
    provider: string;
    endpoint: {
        address: string;
        contract: string;
    };
}
