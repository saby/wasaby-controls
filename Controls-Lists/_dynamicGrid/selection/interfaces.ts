import * as React from 'react';

export type TItemKey = string;
export type TColumnKey = number;

export type TColumns = Date[];

export interface ISelection {
    rootKey: TColumnKey;
    selectedKeys: TColumnKey[];
    firstColumnKey: TColumnKey;
    lastColumnKey: TColumnKey;

    hasBorderTop: boolean;
    hasBorderBottom: boolean;
    gridColumnStart: number | undefined;
    gridColumnEnd: number | undefined;
}

export type TRowSelection = Record<TColumnKey, ISelection>;

export type TGridSelection = Record<
    TItemKey,
    {
        rowSelection: TRowSelection;
        hasSibling: Record<
            TColumnKey,
            {
                up: boolean;
                down: boolean;
            }
        >;
    }
>;

export interface IContextWithSelfRef<T> {
    contextRefForHandlersOnly: React.MutableRefObject<T>;
}
