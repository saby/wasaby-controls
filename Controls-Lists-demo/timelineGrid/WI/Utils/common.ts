import { TNavigationDirection } from 'Controls/interface';
import { Quantum } from 'Controls-Lists/timelineGrid';

export function isWeekend(date: Date): boolean {
    const day = date.getDay();
    const SATURDAY = 6;
    const SUNDAY = 0;
    return day === SATURDAY || day === SUNDAY;
}

export function shiftDate(date: Date, direction: TNavigationDirection, quantum: Quantum): void {
    const shiftSize = direction === 'backward' ? -1 : 1;
    switch (quantum) {
        case Quantum.Second:
        case Quantum.Minute:
        case Quantum.Hour:
            date.setHours(date.getHours() + shiftSize);
            break;
        case Quantum.Day:
            date.setDate(date.getDate() + shiftSize);
            break;
        case Quantum.Month:
            date.setMonth(date.getMonth() + shiftSize);
            break;
    }
}
