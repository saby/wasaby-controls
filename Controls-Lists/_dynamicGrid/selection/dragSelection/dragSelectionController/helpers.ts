import { TPoint, TRect } from './types';
import { TResizerDirection } from '../shared/types';
import { START_FIXED_PART_OBSERVER, END_FIXED_PART_OBSERVER } from '../../../shared/constants';
import { TKeyPair } from './../../shared/types';
import * as React from 'react';

const PERCENT_OF_SIZE_TO_SELECT_CELL = 0.1;

export function getDragStartParams(
    container: HTMLElement,
    topLeftRect: TRect,
    bottomRightRect: TRect,
    direction: TResizerDirection
): {
    startPoint: TPoint;
    xOffset: number;
    yOffset: number;
    xMinOffset: number;
    yMinOffset: number;
} {
    const viewPortRect = getViewportBoundingClientRect(container);
    let xOffset: number;
    let yOffset: number;
    let xMinOffset: number;
    let yMinOffset: number;

    const startPoint: TPoint = {
        x: undefined,
        y: undefined,
    };

    const getXInViewportRange = (x: number) =>
        Math.min(Math.max(x, viewPortRect.x), viewPortRect.x + viewPortRect.width);
    const getYInViewportRange = (y: number) =>
        Math.min(Math.max(y, viewPortRect.y), viewPortRect.y + viewPortRect.height);

    const getXOffsetInViewportRange = (startX: number, offset: number) => {
        // Итоговая координата может быть уменьшена, если выделение вылезает за вьюпорт.
        const correctStartX = getXInViewportRange(startX + offset);
        // Реальный отступ, будет такой же или меньше.
        return offset > 0 ? correctStartX - startX : -1 * (startX - correctStartX);
    };

    const getYOffsetInViewportRange = (startY: number, offset: number) => {
        // Итоговая координата может быть уменьшена, если выделение вылезает за вьюпорт.
        const correctStartY = getYInViewportRange(startY + offset);
        // Реальный отступ, будет такой же или меньше.
        return offset > 0 ? correctStartY - startY : -1 * (startY - correctStartY);
    };

    // Координата Y
    // Начальная точка это либо левый верхний угол, либо правый нижний, но не больше чем размеры контейнера.
    switch (direction) {
        case 'top':
        case 'left':
        case 'topLeft':
        case 'topRight':
            // Правый нижний, центр вокруг которого будет выделение.
            startPoint.y = getYInViewportRange(bottomRightRect.y + bottomRightRect.height);
            // Из левого верхнего вычитаем центр, чтобы узнать отступ(должен быть отрицательным).
            yOffset = getYOffsetInViewportRange(startPoint.y, topLeftRect.y - startPoint.y);
            break;
        case 'bottom':
        case 'right':
        case 'bottomLeft':
        case 'bottomRight':
            // Левый верхний угол, центр вокруг которого будет выделение.
            startPoint.y = getYInViewportRange(topLeftRect.y);
            // Из правого нижнего вычитаем центр, чтобы узнать отступ(должен быть положительным).
            yOffset = getYOffsetInViewportRange(
                startPoint.y,
                bottomRightRect.y + bottomRightRect.height - startPoint.y
            );
            break;
    }

    // Координата X
    // Начальная точка это либо левый верхний угол, либо правый нижний, но не больше чем размеры контейнера.
    switch (direction) {
        case 'top':
        case 'left':
        case 'topLeft':
        case 'bottomLeft':
            // Правый нижний, центр вокруг которого будет выделение.
            startPoint.x = getXInViewportRange(bottomRightRect.x + bottomRightRect.width);
            // Из левого верхнего вычитаем центр, чтобы узнать отступ(должен быть отрицательным).
            xOffset = getXOffsetInViewportRange(startPoint.x, topLeftRect.x - startPoint.x);
            break;
        case 'bottom':
        case 'right':
        case 'topRight':
        case 'bottomRight':
            // Левый верхний угол, центр вокруг которого будет выделение.
            startPoint.x = getXInViewportRange(topLeftRect.x);
            // Из правого нижнего вычитаем центр, чтобы узнать отступ(должен быть положительным).
            xOffset = getXOffsetInViewportRange(
                startPoint.x,
                bottomRightRect.x + bottomRightRect.width - startPoint.x
            );
            break;
    }

    // Минимальное смещение по Y
    switch (direction) {
        case 'top':
        case 'topLeft':
        case 'topRight':
            yMinOffset = -1 * bottomRightRect.height;
            break;
        case 'bottom':
        case 'bottomLeft':
        case 'bottomRight':
            yMinOffset = topLeftRect.height;
            break;
        case 'left':
        case 'right':
            yMinOffset = yOffset;
            break;
    }

    // Минимальное смещение по X
    switch (direction) {
        case 'left':
        case 'topLeft':
        case 'bottomLeft':
            xMinOffset = -1 * bottomRightRect.width;
            break;
        case 'right':
        case 'topRight':
        case 'bottomRight':
            xMinOffset = topLeftRect.width;
            break;
        case 'top':
        case 'bottom':
            xMinOffset = xOffset;
            break;
    }

    return {
        startPoint,
        xOffset,
        yOffset,
        xMinOffset,
        yMinOffset,
    };
}

export function getNewOffset(
    cellRect: TRect,
    startPoint: TPoint,
    direction: TResizerDirection,
    minXOffset: number,
    minYOffset: number
): {
    x: number;
    y: number;
} {
    let newOffsetX: number;
    let newOffsetY: number;

    // Координата Y
    switch (direction) {
        case 'top':
        case 'topLeft':
        case 'topRight':
            newOffsetY = cellRect.y - startPoint.y;
            newOffsetY = newOffsetY > 0 ? minYOffset : newOffsetY;
            break;
        case 'bottom':
        case 'bottomLeft':
        case 'bottomRight':
            newOffsetY = cellRect.y + cellRect.height - startPoint.y;
            newOffsetY = newOffsetY < 0 ? minYOffset : newOffsetY;
            break;
        case 'left':
        case 'right':
            newOffsetY = minYOffset;
            break;
    }

    // Координата X
    switch (direction) {
        case 'left':
        case 'topLeft':
        case 'bottomLeft':
            newOffsetX = cellRect.x - startPoint.x;
            newOffsetX = newOffsetX > 0 ? minXOffset : newOffsetX;
            break;
        case 'right':
        case 'topRight':
        case 'bottomRight':
            newOffsetX = cellRect.x + cellRect.width - startPoint.x;
            newOffsetX = newOffsetX < 0 ? minXOffset : newOffsetX;
            break;
        case 'top':
        case 'bottom':
            newOffsetX = minXOffset;
            break;
    }

    return {
        x: newOffsetX,
        y: newOffsetY,
    };
}

export function getEndPoint(
    startPoint: TPoint,
    direction: TResizerDirection,
    offsetX: number,
    offsetY: number
): TPoint {
    let x: number;
    let y: number;

    // Координата Y
    switch (direction) {
        case 'top':
        case 'left':
        case 'topLeft':
        case 'topRight':
            y = startPoint.y - -1 * offsetY + 2;
            break;
        case 'bottom':
        case 'right':
        case 'bottomLeft':
        case 'bottomRight':
            y = startPoint.y + offsetY - 2;
            break;
    }

    // Координата X
    switch (direction) {
        case 'top':
        case 'left':
        case 'topLeft':
        case 'bottomLeft':
            x = startPoint.x - -1 * offsetX + 2;
            break;
        case 'bottom':
        case 'right':
        case 'topRight':
        case 'bottomRight':
            x = startPoint.x + offsetX - 2;
            break;
    }

    return {
        x,
        y,
    };
}

type TKeyPairs = {
    topLeft: TKeyPair;
    bottomRight: TKeyPair;
};

export function getDragEndParams(
    startKeyPairs: TKeyPairs,
    endKeyPair: TKeyPair,
    direction: TResizerDirection
): TKeyPairs {
    const result: TKeyPairs = {
        topLeft: {
            itemKey: undefined,
            columnKey: undefined,
        },
        bottomRight: {
            itemKey: undefined,
            columnKey: undefined,
        },
    };
    switch (direction) {
        case 'top': {
            result.topLeft.itemKey = endKeyPair.itemKey;
            result.topLeft.columnKey = startKeyPairs.topLeft.columnKey;
            result.bottomRight.itemKey = startKeyPairs.bottomRight.itemKey;
            result.bottomRight.columnKey = startKeyPairs.bottomRight.columnKey;
            break;
        }
        case 'topLeft': {
            result.topLeft = endKeyPair;
            result.bottomRight = startKeyPairs.bottomRight;
            break;
        }
        case 'topRight': {
            result.topLeft.itemKey = endKeyPair.itemKey;
            result.topLeft.columnKey = startKeyPairs.topLeft.columnKey;
            result.bottomRight.itemKey = startKeyPairs.bottomRight.itemKey;
            result.bottomRight.columnKey = endKeyPair.columnKey;
            break;
        }
        case 'bottom': {
            result.topLeft.itemKey = startKeyPairs.topLeft.itemKey;
            result.topLeft.columnKey = startKeyPairs.topLeft.columnKey;
            result.bottomRight.itemKey = endKeyPair.itemKey;
            result.bottomRight.columnKey = startKeyPairs.bottomRight.columnKey;
            break;
        }
        case 'bottomLeft': {
            result.topLeft.itemKey = startKeyPairs.topLeft.itemKey;
            result.topLeft.columnKey = endKeyPair.columnKey;
            result.bottomRight.itemKey = endKeyPair.itemKey;
            result.bottomRight.columnKey = startKeyPairs.bottomRight.columnKey;
            break;
        }
        case 'bottomRight': {
            result.topLeft = startKeyPairs.topLeft;
            result.bottomRight = endKeyPair;
            break;
        }
        case 'left': {
            result.topLeft = endKeyPair;
            result.bottomRight = startKeyPairs.bottomRight;
            break;
        }
        case 'right': {
            result.topLeft = startKeyPairs.topLeft;
            result.bottomRight = endKeyPair;
            break;
        }
    }

    return result;
}

export function getRectParams(startPoint: TPoint, offsetX: number, offsetY: number) {
    const transform = (startPosition: number, offset: number) => {
        if (offset < 0) {
            return {
                size: offset * -1,
                position: startPosition + offset,
            };
        }
        return {
            size: offset,
            position: startPosition,
        };
    };

    const xParams = transform(startPoint.x, offsetX);
    const yParams = transform(startPoint.y, offsetY);

    return {
        x: xParams.position,
        y: yParams.position,
        width: xParams.size,
        height: yParams.size,
    };
}

function getViewportBoundingClientRect(element: HTMLElement): TRect {
    // TODO: Сделать меньше хардкода, но это самый производительный вариант.
    //  Т.к. dynamicGrid крайне специфичный, то пока оставляю так.
    element.classList.remove('tw-contents');
    const containerRect = element.getBoundingClientRect();
    element.classList.add('tw-contents');

    const startFixedPart = element.querySelector(`.${START_FIXED_PART_OBSERVER}`);
    const endFixedPart = element.querySelector(`.${END_FIXED_PART_OBSERVER}`);

    if (startFixedPart) {
        // Можно верить только размерам фиксированных частей.
        // Их координаты будут неверными, т.к. на них висит transform.
        const startFixedRectWidth = startFixedPart.getBoundingClientRect().width;
        containerRect.x += startFixedRectWidth;
        containerRect.width -= startFixedRectWidth;
    }

    if (endFixedPart) {
        // Можно верить только размерам фиксированных частей.
        // Их координаты будут неверными, т.к. на них висит transform.
        containerRect.width -= endFixedPart.getBoundingClientRect().width;
    }

    return containerRect;
}

/**
 * Проверяем, что ячейку можно выделить драгом.
 * Ячейка должна быть полностью видимой.
 * @param e React.MouseEvent.
 */
export function canSelectCellByDrag(e: React.MouseEvent): boolean {
    const cellRect = (e.target as HTMLDivElement).getBoundingClientRect();
    const xSizeToRich = PERCENT_OF_SIZE_TO_SELECT_CELL * cellRect.width;
    const ySizeToRich = PERCENT_OF_SIZE_TO_SELECT_CELL * cellRect.height;
    const availableRect = {
        xStart: cellRect.x + xSizeToRich,
        xEnd: cellRect.x + cellRect.width - xSizeToRich,
        yStart: cellRect.y + ySizeToRich,
        yEnd: cellRect.y + cellRect.height - ySizeToRich,
    };

    return (
        e.pageX >= availableRect.xStart &&
        e.pageX <= availableRect.xEnd &&
        e.pageY >= availableRect.yStart &&
        e.pageY <= availableRect.yEnd
    );
}
