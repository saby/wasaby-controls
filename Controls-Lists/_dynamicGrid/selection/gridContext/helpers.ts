import { TColumnKey, TGridRowSelection, TGridSelection, TPlainSelection } from '../interfaces';

export function resetHasSiblingToDirection(
    selection: TGridRowSelection,
    direction: 'up' | 'down'
): boolean {
    let isChanged = false;

    Object.keys(selection.hasSibling).forEach((key) => {
        if (selection.hasSibling[key][direction]) {
            selection.hasSibling[key][direction] = false;
            isChanged = true;
        }
    });

    return isChanged;
}

function updateSiblingRowSelectionsToDirectionMutable(
    selection: TGridRowSelection,
    siblingSelection: TGridRowSelection,
    currentSelectionKey: TColumnKey,
    direction: 'up' | 'down'
): boolean {
    let isChanged = false;

    const siblingKey = Object.keys(siblingSelection.rowSelection).find(
        (prevKey) =>
            siblingSelection.rowSelection[prevKey].firstColumnKey ===
                selection.rowSelection[currentSelectionKey].firstColumnKey &&
            siblingSelection.rowSelection[prevKey].lastColumnKey ===
                selection.rowSelection[currentSelectionKey].lastColumnKey
    );

    if (typeof siblingKey !== 'undefined') {
        siblingSelection.hasSibling[siblingKey][direction === 'up' ? 'down' : 'up'] = true;
        selection.hasSibling[currentSelectionKey][direction] = true;
        isChanged = true;
    }

    return isChanged;
}

export function updateHasSiblingMutable(
    selection: TGridRowSelection,
    prevSiblingSelection: TGridRowSelection,
    nextSiblingSelection: TGridRowSelection
): [isPrevChanged: boolean, isNextChanged: boolean] {
    Object.keys(selection.rowSelection).forEach((key) => {
        selection.hasSibling[key] = {
            up: false,
            down: false,
        };
    });

    if (!prevSiblingSelection && !nextSiblingSelection) {
        return [false, false];
    }

    let isPrevChanged = false;
    let isNextChanged = false;

    if (prevSiblingSelection) {
        isPrevChanged = resetHasSiblingToDirection(prevSiblingSelection, 'down');
    }

    if (nextSiblingSelection) {
        isNextChanged = resetHasSiblingToDirection(nextSiblingSelection, 'up');
    }

    Object.keys(selection.rowSelection).forEach((currentSelectionKey) => {
        if (prevSiblingSelection) {
            isPrevChanged =
                updateSiblingRowSelectionsToDirectionMutable(
                    selection,
                    prevSiblingSelection,
                    currentSelectionKey as unknown as TColumnKey,
                    'up'
                ) || isPrevChanged;
        }

        if (nextSiblingSelection) {
            isNextChanged =
                updateSiblingRowSelectionsToDirectionMutable(
                    selection,
                    nextSiblingSelection,
                    currentSelectionKey as unknown as TColumnKey,
                    'down'
                ) || isNextChanged;
        }
    });

    if (isPrevChanged) {
        prevSiblingSelection.hasSibling = { ...prevSiblingSelection.hasSibling };
    }

    if (isNextChanged) {
        nextSiblingSelection.hasSibling = { ...nextSiblingSelection.hasSibling };
    }

    return [isPrevChanged, isNextChanged];
}

export function isEqualPlainSelections(
    selection1: TPlainSelection,
    selection2: TPlainSelection
): boolean {
    return compareKeys(Object.keys(selection1), Object.keys(selection2), (a, b) => {
        return compareKeys(selection1[a], selection2[b]);
    });
}

function compareKeys<T>(
    keys1: T[],
    keys2: T[],
    extraComparator?: (key1: T, key2: T) => boolean
): boolean {
    if (keys1.length !== keys2.length) {
        return false;
    } else if (keys1.length === 0) {
        return true;
    } else {
        const sorted1 = keys1.sort();
        const sorted2 = keys2.sort();

        // Ключи есть, их одинаковое количество и они отсортированы.
        for (let i = 0; i < sorted1.length; i++) {
            if (sorted1[i] !== sorted2[i]) {
                return false;
            } else if (extraComparator && !extraComparator(sorted1[i], sorted2[i])) {
                return false;
            }
        }

        return true;
    }
}

export function convertGridSelectionToPlainRecord(selection: TGridSelection): TPlainSelection {
    return Object.keys(selection).reduce((acc, sKey) => {
        return {
            ...acc,
            [sKey]: Object.keys(selection[sKey].rowSelection).reduce(
                (rowSelectedKeys, rsKey) => [
                    ...rowSelectedKeys,
                    ...selection[sKey].rowSelection[rsKey].selectedKeys,
                ],
                []
            ),
        };
    }, {});
}

export function convertPlainSelectionToGridSelection(selection: TPlainSelection): TGridSelection {
    const gridSelection: TGridSelection = {};
    // TODO: Не реализовано, будет делаться тут, если это вообще требуется.
    // Сейчас смена набора выбранных ячеек прикладным разработчиком через слайс не работает.
    return gridSelection;
}
