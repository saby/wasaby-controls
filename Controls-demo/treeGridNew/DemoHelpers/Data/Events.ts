import { IColumn } from 'Controls/grid';
import { IData } from '../Interface';

export const Events = {
    getColumns: (): IColumn[] => {
        return [
            {
                displayProperty: 'title',
                width: '',
            },
        ];
    },
    getDataBeforeExpanded: (): IData[] => {
        return [
            {
                key: 1,
                title: 'Асинхронная обработка перед разворотом узла',
                parent: null,
                nodeType: true,
            },
            { key: 11, title: 'Лист 1-1', parent: 1, nodeType: null },
            { key: 12, title: 'Лист 1-2', parent: 1, nodeType: null },
            {
                key: 2,
                title: 'Асинхронная обработка перед разворотом узла',
                parent: null,
                nodeType: true,
            },
            { key: 21, title: 'Лист 2-1', parent: 2, nodeType: null },
            { key: 22, title: 'Лист 2-2', parent: 2, nodeType: null },
        ];
    },
    getDataBeforeCollapsed: (): IData[] => {
        return [
            {
                key: 1,
                title: 'Асинхронная обработка перед сворачиванием узла',
                parent: null,
                nodeType: true,
            },
            { key: 11, title: 'Лист 1-1', parent: 1, nodeType: null },
            { key: 12, title: 'Лист 1-2', parent: 1, nodeType: null },
            {
                key: 2,
                title: 'Асинхронная обработка перед сворачиванием узла',
                parent: null,
                nodeType: true,
            },
            { key: 21, title: 'Лист 2-1', parent: 2, nodeType: null },
            { key: 22, title: 'Лист 2-2', parent: 2, nodeType: null },
        ];
    },
    getDataCollapsedItemsChanged: (): IData[] => {
        return [
            {
                key: 1,
                title: 'Этот узел нельзя свернуть',
                parent: null,
                nodeType: true,
            },
            { key: 11, title: 'Лист 1-1', parent: 1, nodeType: null },
            { key: 12, title: 'Лист 1-2', parent: 1, nodeType: null },
            {
                key: 2,
                title: 'Этот узел можно свернуть',
                parent: null,
                nodeType: true,
            },
            { key: 21, title: 'Лист 2-1', parent: 2, nodeType: null },
            { key: 22, title: 'Лист 2-2', parent: 2, nodeType: null },
        ];
    },
    getDataExpandedItemsChanged: (): IData[] => {
        return [
            {
                key: 1,
                title: 'Этот узел нельзя развернуть',
                parent: null,
                nodeType: true,
            },
            { key: 11, title: 'Лист 1-1', parent: 1, nodeType: null },
            { key: 12, title: 'Лист 1-2', parent: 1, nodeType: null },
            {
                key: 2,
                title: 'Этот узел можно развернуть',
                parent: null,
                nodeType: true,
            },
            { key: 21, title: 'Лист 2-1', parent: 2, nodeType: null },
            { key: 22, title: 'Лист 2-2', parent: 2, nodeType: null },
        ];
    },
};
