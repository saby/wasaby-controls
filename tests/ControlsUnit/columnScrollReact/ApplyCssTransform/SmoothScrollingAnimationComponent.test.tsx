/**
 * @jest-environment jsdom
 */
import { unmountComponentAtNode } from 'react-dom';
import { WasabyEvents } from 'UICore/Events';
import { fireEvent, waitFor } from '@testing-library/react';

import { IColumnScrollWidths } from 'Controls/_columnScrollReact/common/interfaces';
import SmoothScrollingAnimationComponent from 'Controls/_columnScrollReact/ApplyCssTransform/SmoothScrollingAnimationComponent';
import { SELECTORS } from '../SelectorsForTests';
import { renderStyleInnerPureComponent } from '../forTestsOnly/helpers';

describe('ControlsUnit/columnScrollReact/ApplyCssTransform/SmoothScrollingAnimationComponent', () => {
    let rootContainer: HTMLDivElement = null;
    let componentContainer: HTMLDivElement = null;

    const WIDTHS: Readonly<IColumnScrollWidths> = {
        startFixedWidth: 300,
        endFixedWidth: 0,
        contentWidth: 1000,
        viewPortWidth: 450,
    };

    const initRootContainer = (): HTMLDivElement => {
        const container = document.createElement('div');
        WasabyEvents.initInstance(container);
        document.body.appendChild(container);
        return container;
    };

    const destroyRootContainer = () => {
        unmountComponentAtNode(rootContainer);
        WasabyEvents.destroyInstance(rootContainer);
        rootContainer.remove();
        rootContainer = null;
    };

    const addTransformedContainerToRootContainer = (): HTMLDivElement => {
        const transformedContainer = document.createElement('div');
        transformedContainer.classList.add(SELECTORS.ROOT_TRANSFORMED_ELEMENT);
        rootContainer.appendChild(transformedContainer);
        return transformedContainer;
    };

    const addComponentContainerToRootContainer = (): HTMLDivElement => {
        const container = document.createElement('div');
        rootContainer.appendChild(container);
        return container;
    };

    const destroyComponentContainer = () => {
        unmountComponentAtNode(componentContainer);
        componentContainer.remove();
        componentContainer = null;
    };

    beforeEach(() => {
        rootContainer = initRootContainer();
        componentContainer = addComponentContainerToRootContainer();
    });

    afterEach(() => {
        destroyComponentContainer();
        destroyRootContainer();
    });

    it('no root div in document.', () => {
        expect(() => {
            renderStyleInnerPureComponent(
                <SmoothScrollingAnimationComponent selectors={SELECTORS} {...WIDTHS} from={-10} to={10} />,
                { container: componentContainer }
            );
        }).not.toThrow();
    });

    it('no onAnimationEnd handler.', () => {
        const transformedElement = addTransformedContainerToRootContainer();
        expect(() => {
            renderStyleInnerPureComponent(
                <SmoothScrollingAnimationComponent selectors={SELECTORS} {...WIDTHS} from={100.21} to={-1} />,
                { container: componentContainer }
            );
        }).not.toThrow();
        expect(transformedElement.onanimationend).not.toBeDefined();
    });

    it('base animation style.', () => {
        const { getStyleTag } = renderStyleInnerPureComponent(
            <SmoothScrollingAnimationComponent selectors={SELECTORS} {...WIDTHS} from={0} to={139} />,
            { container: componentContainer }
        );
        expect(getStyleTag()).toMatchSnapshot();
    });

    it('call callback from props onAnimationEnd.', async () => {
        expect.hasAssertions();
        const transformedElement = addTransformedContainerToRootContainer();
        const onAnimationEndMock = jest.fn();

        renderStyleInnerPureComponent(
            <SmoothScrollingAnimationComponent
                selectors={SELECTORS}
                {...WIDTHS}
                from={100.21}
                to={-1}
                onAnimationEnd={onAnimationEndMock}
            />,
            { container: componentContainer }
        );

        fireEvent.animationEnd(transformedElement);

        await waitFor(() => {
            expect(onAnimationEndMock).toBeCalledTimes(1);
        });

        // Завершение другой анимации не вызовет повторный вызов колбека, т.к. мы отписались.
        fireEvent.animationEnd(transformedElement);

        await waitFor(() => {
            expect(onAnimationEndMock).toBeCalledTimes(1);
        });
    });
});
