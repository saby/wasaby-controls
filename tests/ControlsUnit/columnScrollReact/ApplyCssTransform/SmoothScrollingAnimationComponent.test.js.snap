// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlsUnit/columnScrollReact/ApplyCssTransform/SmoothScrollingAnimationComponent base animation style. 1`] = `
<style
  data-qa="ColumnScrollReact__smooth-styles"
>
  @keyframes scrollAnimation {from{transform:translate3d(0px,0,0);}to{transform:translate3d(-139px,0,0);}}@keyframes fixationAnimation {from{transform:translate3d(0px,0,0);}to{transform:translate3d(139px,0,0);}}.js-controls-ColumnScrollReact_rootElement_guid123 { animation-name: scrollAnimation;animation-duration: 0.25s; }.js-controls-ColumnScrollReact_fixed_guid123 { animation-name: fixationAnimation;animation-duration: 0.25s; }
</style>
`;
